src/System.Data.Common/src/System/Data/Common/DbDataReaderExtensions.cs

```diff
@@ -2,22 +2,92 @@
+        {
+            private DataColumnCollection _schemaColumns;
+            private DataRow _schemaRow;
+
+            public DataRowDbColumn(DataRow readerSchemaRow, DataColumnCollection readerSchemaColumns)
+            {
+                _schemaRow = readerSchemaRow;
+                _schemaColumns = readerSchemaColumns;
+                PopulateFields();
+            }
+
+            private void PopulateFields()
+            {
+                AllowDBNull = GetDbColumnValue<bool?>(SchemaTableColumn.AllowDBNull);
+                BaseCatalogName = GetDbColumnValue<string>(SchemaTableOptionalColumn.BaseCatalogName);
+                BaseColumnName = GetDbColumnValue<string>(SchemaTableColumn.BaseColumnName);
+                BaseSchemaName = GetDbColumnValue<string>(SchemaTableColumn.BaseSchemaName);
+                BaseServerName = GetDbColumnValue<string>(SchemaTableOptionalColumn.BaseServerName);
+                BaseTableName = GetDbColumnValue<string>(SchemaTableColumn.BaseTableName);
+                ColumnName = GetDbColumnValue<string>(SchemaTableColumn.ColumnName);
+                ColumnOrdinal = GetDbColumnValue<int?>(SchemaTableColumn.ColumnOrdinal);
+                ColumnSize = GetDbColumnValue<int?>(SchemaTableColumn.ColumnSize);
+                IsAliased = GetDbColumnValue<bool?>(SchemaTableColumn.IsAliased);
+                IsAutoIncrement = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsAutoIncrement);
+                IsExpression = GetDbColumnValue<bool>(SchemaTableColumn.IsExpression);
+                IsHidden = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsHidden);
+                IsIdentity = GetDbColumnValue<bool?>("IsIdentity");
+                IsKey = GetDbColumnValue<bool?>(SchemaTableColumn.IsKey);
+                IsLong = GetDbColumnValue<bool?>(SchemaTableColumn.IsLong);
+                IsReadOnly = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsReadOnly);
+                IsUnique = GetDbColumnValue<bool?>(SchemaTableColumn.IsUnique);
+                NumericPrecision = GetDbColumnValue<int?>(SchemaTableColumn.NumericPrecision);
+                NumericScale = GetDbColumnValue<int?>(SchemaTableColumn.NumericScale);
+                UdtAssemblyQualifiedName = GetDbColumnValue<string>("UdtAssemblyQualifiedName");
+                DataType = GetDbColumnValue<Type>(SchemaTableColumn.DataType);
+                DataTypeName = GetDbColumnValue<string>("DataTypeName");
+            }
+
+            private T GetDbColumnValue<T>(string columnName)
+            {
+                if (_schemaColumns.Contains(columnName) && _schemaRow[columnName] is T value)
+                {
+                    return value;
+                }
+                return default;
+            }
+        }
+
+        public static ReadOnlyCollection<DbColumn> GetColumnSchema(this DbDataReader reader)
         {
-            if (reader.CanGetColumnSchema())
+            if (reader is null)
```

Comment:
nit: to switch. You can even use C# 8 switch expressions to turn this into an expression-bodied method.

Reply:
I mentioned above that i had tried this:

```csharp
public static ReadOnlyCollection<DbColumn> GetColumnSchema2(this DbDataReader reader)
    => reader switch
{
    null => throw new ArgumentNullException(nameof(reader)),
    IDbColumnSchemaGenerator schemaGenerator => schemaGenerator.GetColumnSchema(),
    _ => GetColumnSchemaCompatibility(reader)
};
```

Given that there's no performance benefit the brevity doesn't add anything useful and lessens readability quite a lot imo. I'm not opposed to using c#8 features I just think it's worse if we do in this case. If you feel strongly I can change it but I'd prefer not to.

Reply:
I do find that (slightly) better, but no strong feelings either - it's your call, we can leave as-is.

Reply:
Lets see if anyone else wants to weigh in and I'll go with majority opinion of we get one, if no-one comments I'd like to go with the current version.

Reply:
I am fine with taking this the way @Wraith2 wrote it :smile: